src/util/dictionary.cpp

```diff
@@ -0,0 +1,358 @@
+//  SuperTux
+//  Copyright (C) 2016 Christian Hagemeier <christian@hagemeier.ch>
+//
+//  This program is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU General Public License for more details.
+//
+//  You should have received a copy of the GNU General Public License
+//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#include "util/dictionary.hpp"
+#include <iostream>
+
+using std::vector;
+using std::cout;
+using namespace boost;
+using std::auto_ptr;
+int
+dictionary::hash(std::string key)
+{
+  return ((int) key[0])%10;
+}
+
+dictionary::dictionary(const std::string& filename):
+m_table(),
+m_filename(),
+m_writer()
+{
+  m_filename = filename;
+  // Load from file if it exists
+  this->load();
+}
+
+dictionary::dictionary(void):
+m_table(),
+m_filename(),
+m_writer()
+{
+}
+
+dictionaryTypes dictionary::str2type(std::string type)
+{
+  if(type == "int")
+    return DICT_INT;
+  if(type == "arr")
+    return DICT_ARR;
+  if(type == "bool")
+    return DICT_BOOL;
+  if(type == "float")
+    return DICT_FLOAT;
+  if(type == "table")
+    return DICT_TABLE;
+  if(type == "str")
+    return DICT_STRING;
+  throw new std::runtime_error("No valid type string provided! (Invalid supertux storage file)");
+}
+
+void dictionary::load()
+{
+  try {
+  if(!PHYSFS_exists(m_filename.c_str()))
+  {
+    log_info << m_filename << " doesn't exist, not loading world storage" << std::endl;
+  }
+  else
+  {
+    PHYSFS_Stat statbuf;
+    PHYSFS_stat(m_filename.c_str(), &statbuf);
+    if(statbuf.filetype == PHYSFS_FILETYPE_DIRECTORY)
+    {
+      log_info << m_filename << " is a directory, not loading world storage" << std::endl;
+      return;
+    }
+    auto doc = ReaderDocument::parse(m_filename);
+    auto docroot = doc.get_root();
+
+    if(!(docroot.get_name() == "save-data"))
+    {
+      log_fatal << "file is not a supertux-storage file" << std::endl;
+      throw std::runtime_error("file is not a supertux-storage file");
+    }
+    else
+    {
+      auto mapping = docroot.get_mapping();
+      ReaderCollection items;
+      if(!mapping.get("items",items))
+      {
+        log_fatal << "Ill formed save-data file";
+        throw std::runtime_error("Ill formed save-data file.");
+      }
+      for(ReaderObject item:items.get_objects())
```

Comment:
for const auto& item : items.get_object();

If constis not possible, then at least mark this auto&